### ISSUE TITLE
Bug in suggest.hyperopt, self.metric should be self._metric

### ISSUE THREAD

**Issue by Rassibassi** [ray-project/ray#18547]
Got the following error:
```
File "/home/ME/miniconda3/envs/ray/lib/python3.7/site-packages/ray/tune/suggest/hyperopt.py", line 339, in _to_hyperopt_result
    return {"loss": self.metric_op * result[self.metric], "status": "ok"}
KeyError: 'test_loss'
```
Then I had a look into:

https://github.com/ray-project/ray/blob/ac0a153b0628ca0a1ef3d47da15aa5643c6e61b5/python/ray/tune/suggest/hyperopt.py#L344

and it seems `self.metric` is never defined as such but `self._metric` is.

**Comment by krfricke** [comment]
Hi @Rassibassi,

the error here is that your trial does not report a `test_loss` metric. How does your trainable look like (and your call to `tune.report`)?

`self.metric` is defined in `suggestion.py` and points to `self._metric`:

```
    @property
    def metric(self) -> str:
        """The training result objective value attribute."""
        return self._metric
```

**Comment by Rassibassi** [comment]
Hi,
okay, I see. I am using the `pytorch_lightning` integration.

Here is my code that runs tune:
```python
import os
from functools import partial
from pathlib import Path

import numpy as np
import pytorch_lightning as pl
import torch
from hyperopt import hp
from hyperopt.pyll.base import scope
from pytorch_lightning.loggers import TensorBoardLogger

import ray
from ray import tune
from ray.tune import CLIReporter
from ray.tune.schedulers import ASHAScheduler
from ray.tune.integration.pytorch_lightning import TuneReportCallback
from ray.tune.suggest import ConcurrencyLimiter
from ray.tune.suggest.hyperopt import HyperOptSearch

from custom.lightning.data_modules import myData
from custom.lightning.modules import myDnn


def run_training(conf, tune_callbacks, hparams):
    dm = myData(
        conf["mat_file"],
        batch_size=hparams["batch_size"],
        split_vector=conf["split_vector"],
    )

    model = myDnn(hparams)

    logger = TensorBoardLogger(
        conf["tb_logs"], name=conf["name"], default_hp_metric=False
    )
    trainer = pl.Trainer(
        max_epochs=conf["max_epochs"],
        logger=logger,
        num_processes=conf["num_processes"],
        callbacks=tune_callbacks,
        progress_bar_refresh_rate=0,
    )

    trainer.fit(model, dm)
    trainer.test(model, dm)


def main():
    data_root = Path.home() / "data/database"
    mat_file = "xxx.mat"
    mat_file = str(data_root / mat_file)

    callbacks_str = ["validation", "test"]
    callbacks = [
        TuneReportCallback({f"{cbs}_loss": f"{cbs}_loss"}, on=f"{cbs}_end")
        for cbs in callbacks_str
    ]

    reporter = CLIReporter(
        parameter_columns=[
            "learning_rate",
            "batch_size",
            "std_mult",
            "hidden1",
            "hidden2",
        ],
        metric_columns=["training_iteration", "validation_loss", "test_loss"],
    )

    conf = {
        "name": "blub",
        "max_epochs": 200,
        "split_vector": [0.7, 0.15, 0.15],
        "mat_file": str(mat_file),
        "num_processes": 4,
    }

    conf["tb_logs"] = Path.home() / "results/tb_logs"
    conf["tb_logs"].mkdir(parents=True, exist_ok=True)
    conf["tb_logs"] = str(conf["tb_logs"])

    objective = partial(run_training, conf, callbacks)

    # for debugging
    # hparams = {
    #     'learning_rate': 0.001,
    #     'batch_size': 256,
    #     'std_mult': 3.0,
    #     'n_nodes': [256, 128],
    # }

    hparams = {
        "learning_rate": hp.lognormal("learning_rate", np.log(1.0e-3), 2.0),
        "batch_size": scope.int(
            hp.qloguniform("batch_size", np.log(32), np.log(512), q=32)
        ),
        "std_mult": hp.quniform("std_mult", 1, 3, q=1),
        "hidden1": scope.int(hp.qloguniform("hidden1", np.log(64), np.log(512), q=64)),
        "hidden2": scope.int(hp.qloguniform("hidden2", np.log(64), np.log(512), q=64)),
    }

    algo = HyperOptSearch(hparams, metric="test_loss", mode="min")
    algo = ConcurrencyLimiter(algo, max_concurrent=8)
    scheduler = ASHAScheduler(metric="validation_loss", mode="min")

    analysis = tune.run(
        objective,
        search_alg=algo,
        scheduler=scheduler,
        num_samples=200,
        progress_reporter=reporter,
        resources_per_trial={"cpu": conf["num_processes"], "gpu": 0},
        name=conf["name"],
    )

    print(analysis)


if __name__ == "__main__":
    head_node = os.environ["head_node"]
    port = os.environ["port"]
    ray.init(address=f"{head_node}:{port}")
    main()

```

with my lightning module here:
```python
import torch
from torch.nn import Sequential
import torch.nn.functional as F
import pytorch_lightning as pl

from custom.modules import Denorm
from custom.dnn_models import Dnn

class myDnn(pl.LightningModule):
    def __init__(self, hparams):
        super().__init__()
        self.save_hyperparameters(hparams)

        self.input_dim = 40 + 2
        self.output_dim = 40

        parts = []
        dnn_nodes = [self.input_dim, self.hparams.hidden1, self.hparams.hidden2, self.output_dim]
        parts.append(Dnn(dnn_nodes))
        # the model has more stuff here I just removed it :)
        self.model = Sequential(*parts)

    def forward(self, x):
        return self.model(x)

    def _shared_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self(x)
        loss = F.mse_loss(y_hat, y)
        return loss

    def training_step(self, batch, batch_idx):
        return self._shared_step(batch, batch_idx)

    def validation_step(self, batch, batch_idx):
        return self._shared_step(batch, batch_idx)

    def test_step(self, batch, batch_idx):
        return self._shared_step(batch, batch_idx)

    def configure_optimizers(self):
        return torch.optim.Adam(self.parameters(), lr=self.hparams.learning_rate)

    def _stack_mean(self, outputs):
        return torch.stack(outputs).mean()

    def train_epoch_end(self, outputs):
        avg_loss = self._stack_mean(outputs)
        self.log("train_loss", avg_loss)

    def validation_epoch_end(self, outputs):
        avg_loss = self._stack_mean(outputs)
        self.log("validation_loss", avg_loss)
        self.log("hp/validation_loss", avg_loss)

    def test_epoch_end(self, outputs):
        avg_loss = self._stack_mean(outputs)
        self.log("test_loss", avg_loss)
        self.log("hp/test_loss", avg_loss)

    def on_train_start(self):
        self.logger.log_hyperparams(self.hparams, {"hp/validation_loss": 0, "hp/test_loss": 0})
```

**Comment by Rassibassi** [comment]
Hi again,

I found the problem. I was thinking to have both `validation_loss` and `test_loss`, where `validation_loss` is tracked after each epoch and `test_loss` only at the end, only once after fitting.

Then make the search algorithm based on `test_loss` and the scheduler based on `validation_loss`. As the search algorithm should be based on the best test performance, whereas the scheduler should be able to cancel runs which have bad immediate validation performance.

Somehow, that does not work. However, if I just choose both, search algorithm and scheduler, to be based on the `validation_loss` it runs. 

Is my `test_loss` for the search algorithm and `validation_loss` for scheduler setup unreasonable?

Thanks.

```python
callbacks = [TuneReportCallback(on="validation_end")]
...
algo = HyperOptSearch(hparams, metric="validation_loss", mode="min")
algo = ConcurrencyLimiter(algo, max_concurrent=8)
scheduler = ASHAScheduler(metric="validation_loss", mode="min")
```

**Comment by krfricke** [comment]
We generally expect that each call to `tune.report()` reports the same metrics. Thus, choosing different metrics for the searcher and the scheduler is generally supported - but the fact that they come up sequentially and not at the same time leads to the problem here.

Because each call to `tune.report()` will trigger all `on_trial_result` callbacks, this can lead to confusing situations. For instance, if you train your model for 5 steps and then run validation, the internal iteration counter will be at 6 (because the validation call to `tune.report()` will also be counted). 

Another problem that may come up is that we often use the last observed result, e.g. when deciding about stopping or when fetching the best performing trials. If this is then again based on a metric that is not present in the last call, this will lead to problems again.

Preferably we would have a way to consolidate these reported metrics and ship them in one go. One way I could think of would be to have something like a deferred report (`tune.remember()`? `tune.observe()`?) which would store a metric, but only send it back to Ray Tune with the next call to `tune.report()`. These metrics will then not always be present, but we can ensure that the ones reported in `tune.report` will always be present.

cc @amogkam who is familiar with the PTL integration, @richardliaw @Yard1  for general opinion

**Comment by Yard1** [comment]
I quite like the idea of deferring reports. It should be rather straightforward to implement, too. On that note, I feel that after we provide something like this, we should raise exceptions if the keys reported in `tune.report` differ.
